src/tools/wasp_sqlite.cc

```diff
@@ -0,0 +1,564 @@
+    Exec("insert into table_ values ({}, {}, {}, {});", table.index,
+         table.value.table_type.limits.min,
+         OrNull(table.value.table_type.limits.max),
+         static_cast<int>(table.value.table_type.elemtype));
+  }
+}
+
+void Tool::DoMemorySection(LazyMemorySection section) {
+  for (auto memory : enumerate(section.sequence, imported_memory_count)) {
+    Exec("insert into memory values ({}, {}, {});", memory.index,
+         memory.value.memory_type.limits.min,
+         OrNull(memory.value.memory_type.limits.max));
+  }
+}
+
+void Tool::DoGlobalSection(LazyGlobalSection section) {
+  for (auto global : enumerate(section.sequence, imported_global_count)) {
+    Exec("insert into global values ({}, {}, {});", global.index,
+         static_cast<int>(global.value.global_type.valtype),
+         static_cast<int>(global.value.global_type.mut));
+
+    InsertConstantExpression(global.value.init, "global_init", global.index);
+  }
+}
+
+void Tool::DoExportSection(LazyExportSection section) {
+  for (auto export_ : enumerate(section.sequence)) {
+    Exec("insert into export values ({}, {}, \"{}\", {});", export_.index,
+         static_cast<int>(export_.value.kind), export_.value.name,
+         export_.value.index);
+  }
+}
+
+void Tool::DoStartSection(StartSection section) {
+  if (section) {
+    Exec("insert into start values ({});", section->func_index);
+  }
+}
+
+void Tool::DoElementSection(LazyElementSection section) {
+  for (auto segment : enumerate(section.sequence)) {
+    if (segment.value.is_active()) {
+      const auto& active = segment.value.active();
+      Exec("insert into element values ({}, {}, {});", segment.index,
+           static_cast<int>(segment.value.segment_type()), active.table_index);
+
+      InsertConstantExpression(active.offset, "element_offset", segment.index);
+      for (auto func : enumerate(active.init)) {
+        Exec("insert into element_init values ({}, {}, null, {});",
+             segment.index, func.index, func.value);
+      }
+    } else {
+      const auto& passive = segment.value.passive();
+      Exec("insert into element values ({}, {}, null);", segment.index,
+           static_cast<int>(segment.value.segment_type()));
+      for (auto expr : enumerate(passive.init)) {
+        auto instr = expr.value.instruction;
+        auto opcode_val = static_cast<int>(instr.opcode);
+        switch (instr.opcode) {
+          case Opcode::RefNull:
+            Exec("insert into element_init values ({}, {}, {}, null);",
+                 segment.index, expr.index, opcode_val);
+            break;
+
+          case Opcode::RefFunc:
+            Exec("insert into element_init values ({}, {}, {}, {});",
+                 segment.index, expr.index, opcode_val,
+                 instr.index_immediate());
+            break;
+
+          default:
+            break;
+        }
+      }
+    }
+  }
+}
+
+void Tool::DoCodeSection(LazyCodeSection section) {
+  sqlite3_stmt* stmt;
```

Comment:
Presumably this prepared statement is done for instructions due to performance reasons, and the rest of the code should switch over to them for similar reasons eventually

Reply:
Yeah, there should be some kind of abstraction here. But for now, it was easy enough to call the API directly.

Reply:
Oh probably. Was more thinking that it was only done here because instrs were enough of a hotspot to make it worth doing.

In general I like the pattern of introducing abstractions via refactoring. I.e. write the obvious copy-paste code first, then extract the commonalities out. Especially because I don't know a darn thing about e.g. sqlite, so learning more about how it works first is probably a good idea :d

Reply:
Yeah, this ended up reducing `bb.wasm` time from 2.5s to 1s.